crates/sequencer/src/mempool_syncer.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use std::time::Duration;
+
+use backoff::future::retry as retry_backoff;
+use backoff::ExponentialBackoff;
+use jsonrpsee::core::client::SubscriptionClientT;
+use jsonrpsee::rpc_params;
+use jsonrpsee::ws_client::WsClientBuilder;
+use parking_lot::Mutex;
+use reth_tasks::shutdown::GracefulShutdown;
+use sov_db::ledger_db::SequencerLedgerOps;
+use sov_rollup_interface::rpc::MempoolTransactionSignal;
+use tracing::{debug, error, info, instrument};
+
+#[derive(Clone)]
+pub struct MempoolSyncer<DB>
+where
+    DB: SequencerLedgerOps + Clone + Send + Sync + 'static,
+{
+    /// Database for ledger operations
+    ledger_db: DB,
+    /// Buffer for mempool transactions before storing into the ledger db
+    /// Mapping: Tx hash to rlp encoded transaction
+    transactions_buffer: Arc<Mutex<HashMap<Vec<u8>, Vec<u8>>>>,
+    /// Transaction hashes to be removed from mempool ledger db
+    transactions_to_remove_buffer: Arc<Mutex<HashSet<Vec<u8>>>>,
+    /// sequencer websocket endpoint
+    sequencer_ws_endpoint: String,
+}
+
+impl<DB> MempoolSyncer<DB>
+where
+    DB: SequencerLedgerOps + Clone + Send + Sync + 'static,
+{
+    /// Creates a new MempoolSyncer
+    pub fn new(ledger_db: DB, sequencer_ws_endpoint: String) -> Self {
+        Self {
+            ledger_db,
+            transactions_buffer: Arc::new(Mutex::new(HashMap::new())),
+            transactions_to_remove_buffer: Arc::new(Mutex::new(HashSet::new())),
+            sequencer_ws_endpoint,
+        }
+    }
+
+    #[instrument(name = "MempoolSyncer", skip_all)]
+    pub async fn run(self, shutdown_signal: GracefulShutdown) {
+        let this = self.clone();
+        tokio::spawn(async move {
+            // Start the subscription task
+            this.run_subscription_task().await;
```

Comment:
Pass down the shutdown signal to the spawned tasks so that they can be gracefully shutdown